cirq/google/calibration/phased_fsim.py

```diff
@@ -130,3 +140,217 @@ def override_by(self, other: 'PhasedFSimCharacterization') -> 'PhasedFSimCharact
+
+    @classmethod
+    def _from_json_dict_(
+        cls,
+        parameters: List[Tuple[Qid, Qid, PhasedFSimCharacterization]],
+        gate: Gate,
+        **kwargs,
+    ) -> 'PhasedFSimCalibrationResult':
+        """Magic method for the JSON serialization protocol.
+
+        Converts serialized dictionary into a dict suitable for
+        class instantiation."""
+        return cls(cls._create_parameters_dict(parameters), gate)
+
+    def _json_dict_(self) -> Dict[str, Any]:
+        """Magic method for the JSON serialization protocol."""
+        return {
+            'cirq_type': 'PhasedFSimCalibrationResult',
+            'gate': self.gate,
+            'parameters': [(key[0], key[1], self.parameters[key]) for key in self.parameters],
+        }
+
+
+# We have to relax a mypy constraint, see https://github.com/python/mypy/issues/5374
+@dataclasses.dataclass(frozen=True)  # type: ignore
+class PhasedFSimCalibrationRequest(abc.ABC):
+    """Description of the request to characterize PhasedFSimGate.
+
+    Attributes:
+        pairs: Set of qubit pairs to characterize. A single qubit can appear on at most one pair in
+            the set.
+        gate: Gate to characterize for each qubit pair from pairs. This must be a supported gate
+            which can be described cirq.PhasedFSim gate.
+    """
+
+    # TODO: Validate that each pair is unique and non-overlaping with any other pair.
+    pairs: Tuple[Tuple[Qid, Qid], ...]
+    gate: Gate  # Any gate which can be described by cirq.PhasedFSim
+
+    @abc.abstractmethod
+    def to_calibration_layer(self) -> CalibrationLayer:
+        """Encodes this characterization request in a CalibrationLayer object."""
+
+    @abc.abstractmethod
+    def parse_result(self, result: CalibrationResult) -> PhasedFSimCalibrationResult:
+        """Decodes the characterization result issued for this request."""
+
+
+@json_serializable_dataclass(frozen=True)
+class FloquetPhasedFSimCalibrationOptions:
+    """Options specific to Floquet PhasedFSimCalibration.
+
+    Some angles require another angle to be characterized first so result might have more angles
+    characterized than requested here.
+
+    Attributes:
+        characterize_theta: Whether to characterize θ angle.
+        characterize_zeta: Whether to characterize ζ angle.
+        characterize_chi: Whether to characterize χ angle.
+        characterize_gamma: Whether to characterize γ angle.
+        characterize_phi: Whether to characterize φ angle.
+    """
+
+    characterize_theta: bool
+    characterize_zeta: bool
+    characterize_chi: bool
+    characterize_gamma: bool
+    characterize_phi: bool
+
+
+@dataclasses.dataclass(frozen=True)
+class FloquetPhasedFSimCalibrationResult(PhasedFSimCalibrationResult):
+    """PhasedFSim characterization result specific to Floquet calibration.
+
+    Attributes:
+        options: Options of the characterization from the request.
+    """
+
+    options: FloquetPhasedFSimCalibrationOptions
```

Comment:
Can you explain why we're saving `options` in the `...Result` as well? I understand why this is essential for the `...Request`. Is this just to simplify recordkeeping?

Reply:
If the result type can be more general for a multiple calibration requests, it could make swapping between different calibration routines easier. Instead of class hierarchy, consider using something like an `Optional[Any]` `metadata` field if you really want to tag results with extra info

Reply:
Currently options are almost trivial but in the future they'll potentially grow with gauges to manipulate characterization accuracy as well as  server-based calibrations. My intention was to keep all the the necessary context that describes the data with data itself.

As for class hierarchy: my idea was to share everything which is common to all the characterization techniques in a base class. In that way features which rely only on this generic data (like single-qubit phased calibration presented in a follow-up PR) will work out of the box for all the calibrations. On the other hand more sophisticated features (for example engine simulator presented in a follow-up) might use the specialized "metadata" from the derived class. In case of XEB, it would be nice to put the estimated fidelity there. I was thinking about putting more measures related to the estimation accuracy as well (I'm not sure how much it can be shared between different techniques).

Although I'm not a particular fan of "metadata" because of its lack of structure, I guess it's fine to do it like this. We can re-introduce the class structure any time in the future if it would become necessary

Let me know what you think about this?

Reply:
If the intent is to keep the necessary context with the data, that's fine. We'll have to see how that works in practice, though. Should the parent result class have an abstract property `options`, and derived classes have to fill it in with their specific `Options` type? It would make reading the class hierarchy easier, at the least.

Reply:
It might not be so easy to introduce this property with dataclass and inheritance...

Maybe something like this:
 - I'll define PhasedFSimCalibrationOptions abstract class, empty
 - I'll remove FloquetPhasedFSimCalibrationResult so there will be just PhasedFSimCalibrationResult with option field of type PhasedFSimCalibrationOptions, Each calibration will have its own specialization of Options?
 
This way a type of calibration that was performed will be encoded in the options field, not in the class instance type.

Reply:
What do you think? It's not important enough for me to block this PR; although your suggestion sounds very reasonable

Reply:
Done (as per offline discussion).